src/WholeBodyStateDisplay.cpp

```diff
@@ -619,6 +691,43 @@ void WholeBodyStateDisplay::processWholeBodyState() {
       }
     }
 
+    // Center of pressure per contact. Mainly targets surface contacts (relatively meaningless for point contacts)
+    bool active_contact_in_cop = false;
+    if (use_contact_status_in_cop_) {
+      active_contact_in_cop = contact.status == contact.ACTIVE;
+    } else {
+      active_contact_in_cop = for_dir.norm() > force_threshold_;
+    }
+    bool is_contact_6d = false;
+    if (std::abs(contact.wrench.torque.x) > torque_threshold_ || std::abs(contact.wrench.torque.y) > torque_threshold_) {
+      is_contact_6d = true;
+    }
+    if (cop_enable_ && active_contact_in_cop && is_contact_6d) {
+      // NOTE: x component is negative due to right-hand rotation rule
+      Eigen::Vector3d cop_pos = Eigen::Vector3d(-contact.wrench.torque.y / contact.wrench.force.z,
+                                                  contact.wrench.torque.x / contact.wrench.force.z,
+                                                  0.0);  // Origin of frame is already at contact position
+      if (cop_enable_ && std::isfinite(cop_pos(0)) && std::isfinite(cop_pos(1)) &&
+          std::isfinite(cop_pos(2))) {
+        Ogre::Vector3 cop_point(cop_pos(0), cop_pos(1), cop_pos(2));
+        cop_pos_.push_back(cop_point);
+        boost::shared_ptr<PointVisual> cop;
+        cop.reset(new PointVisual(context_->getSceneManager(), scene_node_));
+        cop->setPoint(cop_point);
+        cop->setFramePosition(contact_pos);
+        cop->setFrameOrientation(contact_orientation);
```

Comment:
Added the contact orientation as the frame orientation here. Please let me know if this appropriate, or if something is missing.

Reply:
I will mark this as resolved to keep our discussion in the above thread.